README.md
# datetimeutils

Utilities for handling dates and times, such as
selecting particular days of the week or month,
formatting timestamps as required by RSS feeds, or
converting timestamp representations of other software
(such as MATLAB and Excel) to R. The package is
lightweight (no dependencies, pure R implementations)
and relies only on R's standard classes to represent
dates and times (Date and POSIXt); it aims to provide
efficient implementations, through vectorisation and
the use of R's native numeric representations of
timestamps where possible.

[ [More] ](http://enricoschumann.net/R/packages/datetimeutils/)

## Installing the package

The latest released version is available from
http://enricoschumann.net. In an R session, just type:

    install.packages('datetimeutils', type = 'source',
                     repos = c('http://enricoschumann.net/R', getOption('repos')))


For the latest development version, check out the Git repository and
build it. In a shell (e.g. sh or bash):

    ## first time: cd to directory and ...
    $ git clone https://github.com/enricoschumann/datetimeutils.git

    ## later: cd to directory and ...
    $ git pull

    ## build and install the package
    $ R CMD build datetimeutils
    $ R CMD INSTALL datetimeutils_0.1-0.tar.gz  ## adjust version number

    ## optionally check
    $ R CMD check datetimeutils_0.1-0.tar.gz    ## adjust version number


NEWS
v0.2-7  (2017-08-25)

  o last_weekday now supports arguments 'before' and
    'inclusive'

v0.2-6  (2017-08-17)

  o new function 'end_of_previous_year'

  o fixed: the 'shift' argument in 'end_of_year' had
    been applied to months instead of years

v0.2-5  (2017-08-07)

  o add vignette with examples

v0.2-4  (2017-07-09)

  o update documentation + internal changes

v0.2-3  (2017-07-09)

  o new function 'guess_datetime'

v0.2-2  (2017-06-26)

  o fixed: 'timegrid' had failed when timezone was
    missing

v0.2-1  (2017-06-26)

  o fixed: 'timegrid' had ignored the 'excludeWeekends'
    argument

v0.2-0  (2017-05-08)

  o *backward* *incompatible* *changes*.  for better
    consistency, several functions have been renamed:

      convertDate   => convert_date
      convertTZ     => convert_tz
      dayOfMonth    => "day_of_month<-"
      dayOfMonth    => day_of_month
      endOfMonth    => end_of_month
      endOfPreviousMonth => end_of_previous_month
      endOfYear     => end_of_year
      firstOfMonth  => first_of_month
      firstOfYear   => first_of_year
      isBusinessDay => is_businessday
      isLeapyear    => is_leapyear
      isWeekend     => is_weekend
      lastWeekday   => last_weekday
      nextBDay      => next_bday
      nextBusinessDay => next_businessday
      nthWeekday    => nth_weekday
      prevBDay      => prev_bday
      previousBusinessDay => previous_businessday

v0.1-5  (2016-10-04)

  o add short-hands 'prevBDay' and 'nextBDay' for
    'previousBusinessDay' and 'nextBusinessDay'

v0.1-4  (2016-07-07)

  o fixed: 'nthWeekday' had failed in some cases

  o new function 'isBusinessDay'

v0.1-3  (2016-05-11)

  o new function 'convertTZ'

v0.1-2  (2016-04-29)

  o new function 'firstOfYear'

v0.1-0  (2015-12-09)

  o there is now a public Git repository at
    https://github.com/enricoschumann/datetimeutils


R/functions.R
## -*- truncate-lines: t; -*-

                                        # DATES

prev_bday  <- previous_businessday <- function (x, holidays = NULL, shift = -1) {
    if (!is.null(holidays))
        .NotYetUsed("holidays", FALSE)
    if (!all(inherits(x, "Date") | inherits(x, "POSIXt")))
        stop("input must inherit from classes ",
             sQuote("Date"), " or ", sQuote("POSIXt"))
    x <- as.Date(x)
    if (shift == -1L || shift == 0L) {
        x <- x + shift
        tmp <- as.POSIXlt(x)
        tmpi <- tmp$wday == 6L
        x[tmpi] <- x[tmpi] - 1L
        tmpi <- tmp$wday == 0L
        x[tmpi] <- x[tmpi] - 2L
    } else {
        for (i in 1:(-shift)) {
            x <- x - 1
            tmp <- as.POSIXlt(x)
            tmpi <- tmp$wday == 6L
            x[tmpi] <- x[tmpi] - 1L
            tmpi <- tmp$wday == 0L
            x[tmpi] <- x[tmpi] - 2L
        }
    }
    x
}

next_bday  <- next_businessday <- function(x, holidays = NULL, shift = 1) {
    if (!is.null(holidays))
        .NotYetUsed("holidays", FALSE)
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    x <- as.Date(x)
    if (shift == 1L || shift == 0L) {
        x <- x + shift
        tmp <- as.POSIXlt(x)
        tmpi <- tmp$wday == 6L
        x[tmpi] <- x[tmpi] + 2L
        tmpi <- tmp$wday == 0L
        x[tmpi] <- x[tmpi] + 1L
        
    } else {
        for (i in 1:shift) {
            x <- x + 1
            tmp <- as.POSIXlt(x)
            tmpi <- tmp$wday == 6L
            x[tmpi] <- x[tmpi] + 2L
            tmpi <- tmp$wday == 0L
        x[tmpi] <- x[tmpi] + 1L
        }
    }
    x
}

is_weekend <- function(x) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$wday == 0L | tmp$wday == 6L
}

is_leapyear <- function(x)
    x %% 4 == 0 & (x %% 100 != 0 | x %% 400 == 0)

is_businessday <- function(x, holidays = NULL) {
    if (!is.null(holidays))
        .NotYetUsed("holidays", FALSE)
    !is_weekend(x)
}

first_of_month <- function (x) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$mday <- 1L
    as.Date(tmp)
}

first_of_year <- function (x) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$mday <- 1L
    tmp$mon <- 0L
    as.Date(tmp)
}

end_of_month <- function(x, shift = 0L) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$mon <- tmp$mon + 1L + shift
    tmp$mday <- 1L
    as.Date(tmp) - 1L
}

end_of_year <- function(x, shift = 0L) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$year <- tmp$year + shift
    tmp$mon <- 12
    tmp$mday <- 1L
    as.Date(tmp) - 1L
}

end_of_previous_month <- function(x)
    end_of_month(x, -1L)

end_of_previous_year <- function(x)
    end_of_year(x, shift = -1L)

mday <- function(x)
    day_of_month(x)

`mday<-` <- function(x, value){
    day_of_month(x) <- value
    x
}

day_of_month <- function(x) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    as.POSIXlt(x)$mday
}

`day_of_month<-` <- function(x, value) {
    if (!all(inherits(x, "Date") | inherits(x, "POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    cl <- class(x)
    tmp <- as.POSIXlt(x)
    tmp$mday <- value
    if (cl == "Date")
        as.Date(tmp)
    else if (cl == "POSIXct")
        as.POSIXct(tmp)
    else 
        tmp
}

last_weekday <- function(weekday, x, shift = 0L,
                        period = "month", before, inclusive = TRUE) {
    if (missing(before) && !all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    if (missing(before)) {
        tmp <- as.POSIXlt(x)
        tmp$mon <- tmp$mon + 1L
        tmp$mday <- 1L
        ldate <- as.Date(tmp) - 1L
    } else
        ldate <- if (inclusive) before else before - 1L
    lweekday <- as.POSIXlt(ldate)$wday
    ldate - (lweekday - weekday)%%7L + (shift*7L)
}

nth_weekday <- function(weekday, x, n = 1L) {
    if (!all(inherits(x,"Date") | inherits(x,"POSIXt")))
        stop("input must inherit from class Date or POSIXt")
    tmp <- as.POSIXlt(x)
    tmp$mday <- 1L
    tmp <- as.POSIXlt(as.Date(tmp)) ## 'wday' is not recomputed
    weekday1 <- tmp$wday            ## when 'mday' is changed
    as.Date(tmp) + (weekday - weekday1) %% 7L + 7L*(n - 1L)
}
                                        # TIMES
make_hhmmss <- function(x, label = "time specification (HHMMSS)") {
    x <- as.character(x)
    if (nchar(x) == 1L)
        x <- paste("0", x, "0000", sep = "")
    else if (nchar(x) == 2L)
        x <- paste(x, "0000", sep = "")
    else if (nchar(x) == 4L)
        x <- paste(x, "00", sep = "")

    ss <- substr(x, 1, 2)
    if (ss > "24" || ss < "00")
        stop("check ", label)
    ss <- substr(x, 3, 4)
    if (ss > "60" || ss < "00")
        stop("check ", label)
    ss <- substr(x, 5, 6)
    if (ss > "60" || ss < "00")
        stop("check ", label)
    x
}

timegrid <- function(from, to, interval,
                     excludeWeekends = TRUE,
                     holidays   = NULL,
                     fromHHMMSS = "080000",
                     toHHMMSS   = "220000") {

    if (missing(interval))
        stop(sQuote("interval"), " missing")
    if (!inherits(from, "POSIXt") || !inherits(to, "POSIXt"))
        stop(sQuote("from"), " and ", sQuote("to"),
             " must inherit from POSIXt")
    fromHHMMSS <- make_hhmmss(fromHHMMSS)
    toHHMMSS   <- make_hhmmss(toHHMMSS)
    
    grd <- seq(from, to, by = interval)
    if (!is.null(holidays)) {
        if (!inherits(holidays, "Date"))
            holidays <- as.Date(holidays)
        grd <- grd[!(as.Date(grd) %in% holidays)]
    }
    tz <- attr(from, "tzone")
    lt <- as.POSIXlt(grd, tz = if (is.null(tz)) "" else tz)
    tmp <- lt$hour*10000 + lt$min*100 + lt$sec
    if (excludeWeekends)
        grd <- grd[lt$wday > 0L & lt$wday < 6L &
                   as.numeric(fromHHMMSS) <= tmp &
                   as.numeric(toHHMMSS) >= tmp] 
    as.POSIXct(grd)
}

roundPOSIXt <- function(t, interval, up = FALSE) {

    if (!inherits(t, "POSIXct"))
        t <- as.POSIXct(t)
    tz <- attr(t, "tzone")
    n <- as.integer(strsplit(interval, " ", fixed = TRUE)[[c(1L,1L)]])

    if (grepl("sec", interval, ignore.case = TRUE)) {
        ## Default unit in 'POSIXct' is 'sec': do nothing.
    } else if (grepl("min", interval, ignore.case = TRUE)) {
        n <- 60L * n
    } else if (grepl("hour", interval, ignore.case = TRUE)) {
        n <- 60L * 60L * n
    } else if (grepl("day", interval, ignore.case = TRUE)) {
        n <- 24L * 60L * 60L * n
    } else {
        stop(sQuote("interval"), " not clear")
    }
    if (up)
        .POSIXct(n*ceiling(unclass(t)/n), tz = tz)
    else
        .POSIXct(n*floor(unclass(t)/n), tz = tz)
}

ssm <- function(time, tz = "") {
    if (tz == "UTC" || tz == "GMT")
        as.numeric(time) %% 86400
    else {
        pt <- as.POSIXlt(time, tz)        
        3600 * pt$hour + 60 * pt$min + pt$sec
    }
}

convert_date <- function(x, type, fraction = FALSE, tz = "") {
    type <- tolower(type)
    if (type == "excel" && !fraction){
        as.Date(x, origin = "1899-12-30")
    } else if (type == "excel") {
        tmp <- as.POSIXct(x * 86400, origin = "1899-12-30", tz = "UTC")
        as.POSIXct(strptime(format(tmp), format = "%Y-%m-%d %H:%M:%S", tz = tz))        
    } else if (type == "matlab" && !fraction) {
        as.Date(x, origin = "1970-01-01") - 719529
    } else if (type == "matlab" && fraction) {
        tmp <- as.POSIXct((x - 719529) * 86400, origin = "1970-01-01")
        as.POSIXct(strptime(format(tmp), format = "%Y-%m-%d %H:%M:%S", tz = tz))        
    } else
        stop("unknown type")
}

rfc822t <- function(x, include.dow = TRUE) {

    days <- c("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
    xx <- as.POSIXlt(x)
    if (include.dow)
        paste0(days[xx$wday+1], ", ", xx$mday, " ", month.abb[xx$mon + 1], " ",
               format(xx, "%Y %H:%M:%S %z"))
    else
        paste0(xx$mday, " ", month.abb[xx$mon + 1], " ",
               format(xx, "%Y %H:%M:%S %z"))
}

convert_tz <- function(datetime, from = "", to = "") {
    if (inherits(datetime, "POSIXt"))
        datetime <- format(datetime, tz = from,
                           format = "%Y-%m-%d %H:%M:%S")
    ans <- as.POSIXct(datetime, tz = from)
    attr(ans, "tzone") <- to
    ans
}

## timestamps <- seq(as.Date("2015-1-1"), as.Date("2015-12-15"), by = "1 day")
## ultimo, firstofmonth, 
ref_timestamp <- function(what, when = Sys.Date(), timestamps, index = FALSE) {
    what <- tolower(what)
    if (!is.null(when) && what == "mtd") {
        ii <- suppressWarnings(max(which(as.Date(timestamps) <=
                                         end_of_previous_month(as.Date(when)))))
        ii[ii == -Inf] <- 0
        
    } else if (!is.null(when) && what == "ytd") {
        ii <- suppressWarnings(max(which(as.Date(timestamps) <=
                                         end_of_previous_year(as.Date(when)))))
        ii[ii == -Inf] <- 0
    }
    if (index)
        timestamps[ii] else ii
}



.dt_patterns <- c(
    "[1-2][0-9][0-9][0-9]-[0-9]+-[0-9]+ +[0-9]+:[0-9]+:[0-9]+",           "%Y-%m-%d %H:%M:%S",
    "[0-9][0-9]+-[0-9]+-[0-9]+ +[0-9]+:[0-9]+",                           "%Y-%m-%d %H:%M",
    "[0-9]+/[0-9]+/[0-9][0-9] +[0-9]+:[0-9]+:[0-9]+",                     "%m/%d/%y %H:%M:%S",
    "[0-9]+/[0-9]+/[1-2][0-9][0-9][0-9] +[0-9]+:[0-9]+:[0-9]+",           "%m/%d/%Y %H:%M:%S",
    "[0-9]+/[0-9]+/[0-9][0-9] +[0-9]+:[0-9]+",                            "%m/%d/%y %H:%M",
    "[0-9]+/[0-9]+/[1-2][0-9][0-9][0-9] +[0-9]+:[0-9]+",                  "%m/%d/%Y %H:%M",
    "[0-9][0-9]*[.][0-9]+[.][1-2][0-9][0-9][0-9]* +[0-9]+:[0-9]+:[0-9]+", "%d.%m.%Y %H:%M:%S",
    "[0-9][0-9]*[.][0-9]+[.][1-2][0-9][0-9][0-9]* +[0-9]+:[0-9]+",        "%d.%m.%Y %H:%M"
)

guess_datetime <- function(s) {

    x <- as.character(s)
    ans <- .POSIXct(rep(NA_real_, length(x)))
    done <- logical(length(x))

    ii <- seq(1, length(.dt_patterns), by = 2L)

    for (t in ii) {

        i <- grepl(.dt_patterns[t], x) & !done
        ans[i] <- as.POSIXct(strptime(x[i], .dt_patterns[t + 1L]))
        done[i] <- TRUE

        if (all(done))
            break
    }

    ans
}


inst/unitTests/runTests.R
pkg <- "datetimeutils"
if (tolower(Sys.getenv("ES_PACKAGE_TESTING")) == "true" &&
    require("RUnit", quietly = TRUE) &&
    require(pkg, quietly = TRUE, character.only = TRUE)) {

    path <- paste0("~/Packages/", pkg, "/inst/unitTests")

    ## if the package is installed and you want to run
    ## the test, use this path:    
    ##
    ##     path <- system.file("unitTests", package = "PMwR")
    ##
    
    myTestSuite <- defineTestSuite(pkg,
                                   dirs = path,
                                   testFileRegexp = "ut_.*")
    stopifnot(isValidTestSuite(myTestSuite))
    testResult <- runTestSuite(myTestSuite, verbose = 0L)
    printTextProtocol(testResult,
                      showDetails = TRUE,
                      fileName = paste0(file.path(path, "test_results"), ".txt"))
}


inst/unitTests/ut_datetime.R
test.last_weekday <- function() {

    x <- as.Date(c("1996-05-30","1996-05-20"))

    checkEquals(last_weekday(6, x, 0),
                structure(c(9641, 9641), class = "Date"))

    checkEquals(last_weekday(6, x, -1),
                structure(c(9641, 9641), class = "Date") - 7)

    checkEquals(last_weekday(1:5, as.Date("2017-1-1")),
                last_weekday(1:5, before=as.Date("2017-1-31")))
    checkEquals(last_weekday(1:5, as.Date("2017-2-28")),
                last_weekday(1:5, before=as.Date("2017-2-28")))
                 
    tmp <- seq(from = as.Date("2017-01-01"),
               to   = as.Date("2018-12-31"),
               by   = "1 day")

    for (i in 1:5) {
        checkEquals(last_weekday(i, before = tmp[as.POSIXlt(tmp)$wday == i]),
                    tmp[as.POSIXlt(tmp)$wday == i]-0)
        checkEquals(last_weekday(i, before = tmp[as.POSIXlt(tmp)$wday == i],
                                 inclusive = FALSE),
                    tmp[as.POSIXlt(tmp)$wday == i]-7)
    }    
}

test.convert_date <- function() {
    checkEquals(convert_date(41824, "excel"), as.Date("2014-07-04"))
    checkEquals(convert_date(61, "excel"), as.Date("1900-03-01"))
}

test.roundPOSIXt <- function() {
    tmp <- structure(1435589310.11177,
                     class = c("POSIXct", "POSIXt"),
                     tzone = "GMT")

    checkEquals(roundPOSIXt(tmp, "1 hour"),
                structure(1435586400,
                          tzone = "GMT",
                          class = c("POSIXct", "POSIXt")))

    checkEquals(roundPOSIXt(tmp, "1 hour", TRUE),
                structure(1435590000,
                          tzone = "GMT",
                          class = c("POSIXct", "POSIXt")))    
}


test.next_businessday <- function() {
    dates <- seq(as.Date("2012-1-1"), as.Date("2015-1-10"), by = "1 day")
    should <- dates + 1
    should[as.POSIXlt(dates)$wday == 5L] <- dates[as.POSIXlt(dates)$wday == 5L] + 3
    should[as.POSIXlt(dates)$wday == 6L] <- dates[as.POSIXlt(dates)$wday == 6L] + 2
    should[as.POSIXlt(dates)$wday == 0L] <- dates[as.POSIXlt(dates)$wday == 0L] + 1
    checkTrue(all(should == next_businessday(dates)))

    should <- dates + 0
    should[as.POSIXlt(dates)$wday == 6L] <- dates[as.POSIXlt(dates)$wday == 6L] + 2
    should[as.POSIXlt(dates)$wday == 0L] <- dates[as.POSIXlt(dates)$wday == 0L] + 1
    checkTrue(all(should == next_businessday(dates, shift = 0)))

    checkTrue(all(
        next_businessday(next_businessday(dates)) ==
            next_businessday(dates, shift = 2)))
    checkTrue(all(
        next_bday(next_bday(dates)) == next_bday(dates, shift = 2)))

    checkTrue(all(
        next_businessday(next_businessday(next_businessday(dates))) ==
        next_businessday(dates, shift = 3)))
    checkTrue(all(
        next_bday(next_bday(next_bday(dates))) == next_bday(dates, shift = 3)))
}

test.previous_businessday <- function() {
    dates <- seq(as.Date("2012-1-1"), as.Date("2015-1-10"), by = "1 day")
    should <- dates - 1
    should[as.POSIXlt(dates)$wday == 1L] <- dates[as.POSIXlt(dates)$wday == 1L] - 3
    should[as.POSIXlt(dates)$wday == 6L] <- dates[as.POSIXlt(dates)$wday == 6L] - 1
    should[as.POSIXlt(dates)$wday == 0L] <- dates[as.POSIXlt(dates)$wday == 0L] - 2
    checkTrue(all(should == previous_businessday(dates)))

    should <- dates + 0
    should[as.POSIXlt(dates)$wday == 6L] <- dates[as.POSIXlt(dates)$wday == 6L] - 1
    should[as.POSIXlt(dates)$wday == 0L] <- dates[as.POSIXlt(dates)$wday == 0L] - 2
    checkTrue(all(should == previous_businessday(dates, shift = 0)))

    checkTrue(all(
        previous_businessday(previous_businessday(dates)) ==
            previous_businessday(dates, shift = -2)))
    checkTrue(all(
        prev_bday(prev_bday(dates)) == prev_bday(dates, shift = -2)))

    checkTrue(all(
        previous_businessday(previous_businessday(previous_businessday(dates))) ==
            previous_businessday(dates, shift = -3)))    
    checkTrue(all(
        prev_bday(prev_bday(prev_bday(dates))) == prev_bday(dates, shift = -3)))    
}

test.nth_weekday <- function() {

    checkTrue(all(nth_weekday(1, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-4")))
    checkTrue(all(nth_weekday(2, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-5")))
    checkTrue(all(nth_weekday(3, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-6")))
    checkTrue(all(nth_weekday(4, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-7")))
    checkTrue(all(nth_weekday(5, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-1")))
    checkTrue(all(nth_weekday(6, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-2")))
    checkTrue(all(nth_weekday(7, as.Date("2016-1-1")+0:30, 1) ==
                  as.Date("2016-1-3")))

    ## 2016 is a leap year
    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-1")))
    checkTrue(all(nth_weekday(2, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-2")))
    checkTrue(all(nth_weekday(3, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-3")))
    checkTrue(all(nth_weekday(4, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-4")))
    checkTrue(all(nth_weekday(5, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-5")))
    checkTrue(all(nth_weekday(6, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-6")))
    checkTrue(all(nth_weekday(7, as.Date("2016-2-1")+0:28, 1) ==
                  as.Date("2016-2-7")))

    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 2) ==
                  as.Date("2016-2-8")))
    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 3) ==
                  as.Date("2016-2-15")))
    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 4) ==
                  as.Date("2016-2-22")))
    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 5) ==
                  as.Date("2016-2-29")))

    ## 6th Monday of February == 1st Monday of March
    checkTrue(all(nth_weekday(1, as.Date("2016-2-1")+0:28, 6) ==
                  as.Date("2016-3-7")))    
}

test.timegrid <- function() {

    x <- timegrid(as.POSIXlt("2017-06-23 21:00:00", tz = "UTC"),
                  as.POSIXlt("2017-06-26 10:00:00", tz = "UTC"),
                  interval = "15 min")
    checkEquals(x,
                structure(c(1498251600, 1498252500, 1498253400,
                            1498254300, 1498255200, 1498464000,
                            1498464900, 1498465800, 1498466700,
                            1498467600, 1498468500, 1498469400,
                            1498470300, 1498471200),
                          class = c("POSIXct", "POSIXt"), tzone = "UTC"))

    x <- timegrid(as.POSIXlt("2017-06-23 21:00:00", tz = "Europe/Zurich"),
                  as.POSIXlt("2017-06-26 10:00:00", tz = "Europe/Zurich"),
                  interval = "15 min")
    checkEquals(x,
                structure(c(1498244400, 1498245300, 1498246200,
                            1498247100, 1498248000, 1498456800,
                            1498457700, 1498458600, 1498459500,
                            1498460400, 1498461300, 1498462200,
                            1498463100, 1498464000),
                          class = c("POSIXct", "POSIXt"),
                          tzone = "Europe/Zurich"))

}

test.guess_datetime <- function() {

    s <- c("  1999-08-19     10:00:00   ",
           "1999-08-19 10:00",
           "19.8.1999 10:00",
           "8/19/99  10:00:00",
           "8/19/1999 10:00:00",
           "19.8.1999 10:00:00",
           "8/19/1999    10:00:00   ",
           "19.8.1999    10:00:00   "
           )
    
    checkTrue(all(as.character(guess_datetime(s)) ==
                  "1999-08-19 10:00:00"))

}

test.end_of_year <- function() {
    checkTrue(all(
        end_of_year(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01"))) == as.Date("2017-12-31")))

    checkTrue(all(
        end_of_year(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = 0) == as.Date("2017-12-31")))

    checkTrue(all(
        end_of_year(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = 1) == as.Date("2018-12-31")))

    checkTrue(all(
        end_of_year(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = -1) == as.Date("2016-12-31")))

    checkTrue(all(
        end_of_previous_year(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01"))) == as.Date("2016-12-31")))
}

test.end_of_month <- function() {
    checkTrue(all(
        end_of_month(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01"))) ==
        as.Date(
            c("2017-01-31",
              "2017-12-31",
              "2017-08-31"))))

    checkTrue(all(
        end_of_month(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = 0) ==
        as.Date(
            c("2017-01-31",
              "2017-12-31",
              "2017-08-31"))))
    
    checkTrue(all(
        end_of_month(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = 1) ==
        as.Date(
            c("2017-02-28",
              "2018-01-31",
              "2017-09-30"))))

    checkTrue(all(
        end_of_month(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01")), shift = -1) ==
        as.Date(
            c("2016-12-31",
              "2017-11-30",
              "2017-07-31"))))

    checkTrue(all(
        end_of_previous_month(as.Date(
            c("2017-01-01",
              "2017-12-31",
              "2017-08-01"))) ==
        as.Date(
            c("2016-12-31",
              "2017-11-30",
              "2017-07-31"))))    
}
